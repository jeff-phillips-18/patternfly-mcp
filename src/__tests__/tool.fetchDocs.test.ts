import { McpError } from '@modelcontextprotocol/sdk/types.js';
import { fetchDocsTool } from '../tool.fetchDocs';
import { processDocsFunction } from '../server.getResources';

// Mock dependencies
jest.mock('../server.getResources');
jest.mock('../server.caching', () => ({
  memo: jest.fn(fn => fn)
}));

const mockProcessDocs = processDocsFunction as jest.MockedFunction<typeof processDocsFunction>;

describe('fetchDocsTool', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should have a consistent return structure', () => {
    const tool = fetchDocsTool();

    expect(tool).toMatchSnapshot('structure');
  });
});

describe('fetchDocsTool, callback', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it.each([
    {
      description: 'default',
      value: 'components/button.md',
      urlList: ['components/button.md']
    },
    {
      description: 'multiple files',
      value: 'combined docs content',
      urlList: ['components/button.md', 'components/card.md', 'components/table.md']
    },
    {
      description: 'with empty files',
      value: 'trimmed content',
      urlList: ['components/button.md', '', '   ', 'components/card.md', 'components/table.md']
    },
    {
      description: 'with empty urlList',
      value: 'empty content',
      urlList: []
    },
    {
      description: 'with empty strings in a urlList',
      value: 'trimmed and empty content',
      urlList: ['', ' ']
    },
    {
      description: 'with invalid urlList',
      value: 'invalid path',
      urlList: ['invalid-url']
    }
  ])('should parse parameters, $description', async ({ value, urlList }) => {
    mockProcessDocs.mockResolvedValue(value);
    const [_name, _schema, callback] = fetchDocsTool();
    const result = await callback({ urlList });

    expect(mockProcessDocs).toHaveBeenCalledWith(urlList);
    expect(result).toMatchSnapshot();
  });

  it.each([
    {
      description: 'with missing or undefined urlList',
      error: 'Missing required parameter: urlList',
      urlList: undefined
    },
    {
      description: 'with null urlList',
      error: 'Missing required parameter: urlList',
      urlList: null
    },
    {
      description: 'when urlList is not an array',
      error: 'must be an array of strings',
      urlList: 'not-an-array'
    }
  ])('should handle errors, $description', async ({ error, urlList }) => {
    const [_name, _schema, callback] = fetchDocsTool();

    await expect(callback({ urlList })).rejects.toThrow(McpError);
    await expect(callback({ urlList })).rejects.toThrow(error);
  });

  it('should handle processing errors', async () => {
    mockProcessDocs.mockRejectedValue(new Error('Network error'));
    const [_name, _schema, callback] = fetchDocsTool();

    await expect(callback({ urlList: ['missing.md'] })).rejects.toThrow(McpError);
    await expect(callback({ urlList: ['missing.md'] })).rejects.toThrow('Failed to fetch documentation');
  });
});
