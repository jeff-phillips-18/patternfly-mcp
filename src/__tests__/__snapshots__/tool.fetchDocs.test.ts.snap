// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`fetchDocsTool should have a consistent return structure: structure 1`] = `
[
  "fetchDocs",
  {
    "description": "Fetch documentation for one or more URLs extracted from previous tool calls responses. The URLs should be passed as an array in the "urlList" argument.",
    "inputSchema": {
      "urlList": ZodArray {
        "_def": {
          "description": "The list of URLs to fetch documentation from",
          "exactLength": null,
          "maxLength": null,
          "minLength": null,
          "type": ZodString {
            "_def": {
              "checks": [],
              "coerce": false,
              "typeName": "ZodString",
            },
            "and": [Function],
            "array": [Function],
            "brand": [Function],
            "catch": [Function],
            "default": [Function],
            "describe": [Function],
            "isNullable": [Function],
            "isOptional": [Function],
            "nullable": [Function],
            "nullish": [Function],
            "optional": [Function],
            "or": [Function],
            "parse": [Function],
            "parseAsync": [Function],
            "pipe": [Function],
            "promise": [Function],
            "readonly": [Function],
            "refine": [Function],
            "refinement": [Function],
            "safeParse": [Function],
            "safeParseAsync": [Function],
            "spa": [Function],
            "superRefine": [Function],
            "transform": [Function],
            "~standard": {
              "validate": [Function],
              "vendor": "zod",
              "version": 1,
            },
          },
          "typeName": "ZodArray",
        },
        "and": [Function],
        "array": [Function],
        "brand": [Function],
        "catch": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "pipe": [Function],
        "promise": [Function],
        "readonly": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
    },
  },
  [Function],
]
`;

exports[`fetchDocsTool, callback should parse parameters, default 1`] = `
{
  "content": [
    {
      "text": "components/button.md",
      "type": "text",
    },
  ],
}
`;

exports[`fetchDocsTool, callback should parse parameters, multiple files 1`] = `
{
  "content": [
    {
      "text": "combined docs content",
      "type": "text",
    },
  ],
}
`;

exports[`fetchDocsTool, callback should parse parameters, with empty files 1`] = `
{
  "content": [
    {
      "text": "trimmed content",
      "type": "text",
    },
  ],
}
`;

exports[`fetchDocsTool, callback should parse parameters, with empty strings in a urlList 1`] = `
{
  "content": [
    {
      "text": "trimmed and empty content",
      "type": "text",
    },
  ],
}
`;

exports[`fetchDocsTool, callback should parse parameters, with empty urlList 1`] = `
{
  "content": [
    {
      "text": "empty content",
      "type": "text",
    },
  ],
}
`;

exports[`fetchDocsTool, callback should parse parameters, with invalid urlList 1`] = `
{
  "content": [
    {
      "text": "invalid path",
      "type": "text",
    },
  ],
}
`;
